components/script_bindings/codegen/codegen.py

```diff
@@ -944,39 +944,74 @@ def fromJSValTemplate(config, errorHandler, exceptionCode):
 
     if is_typed_array(type):
         if failureCode is None:
-            unwrapFailureCode = (f'throw_type_error(*cx, "{sourceDescription} is not a typed array.");\n'
-                                 f'{exceptionCode}')
+            unwrapFailureCode = (
+                f'throw_type_error(*cx, "{sourceDescription} is not a typed array.");\n'
+                f'{exceptionCode}'
+            )
         else:
             unwrapFailureCode = failureCode
 
-        typeName = type.unroll().name  # unroll because it may be nullable
+        # Unroll nullable types to get the base type name
+        typeName = type.unroll().name
 
-        if isMember == "Union":
+        # For unions, we must use the Heap variant (HeapArrayBuffer, HeapArrayBufferView, etc.)
+        # because only the Heap version implements `JSTraceable` and can safely be rooted.
+        is_union_member = (isMember == "Union")
+        if is_union_member:
             typeName = f"Heap{typeName}"
 
-        templateBody = fill(
-            """
-            match typedarray::${ty}::from($${val}.get().to_object()) {
-                Ok(val) => val,
-                Err(()) => {
-                    $*{failureCode}
+        if is_union_member:
```

Comment:
Let's merge this with the previous `if is_union_member`

Reply:
I think this was not addressed yet.